cleanrl/sac_atari.py

```diff
@@ -0,0 +1,359 @@
+            nn.ReLU(),
+            nn.Conv2d(64, 64, kernel_size=3, stride=1),
+            nn.Flatten(),
+            nn.ReLU(),
+        )
+
+        with torch.inference_mode():
+            output_dim = self.conv(torch.zeros(1, *obs_shape)).shape[1]
+
+        self.fc_logits = nn.Sequential(
+            nn.Linear(output_dim, 512),
+            nn.ReLU(),
+            nn.Linear(512, envs.single_action_space.n),
+        )
+
+    def forward(self, x):
+        x = self.conv(x)
+        logits = self.fc_logits(x)
+
+        return logits
+
+    def get_action(self, x):
+        logits = self(x / 255.0)
+        policy_dist = Categorical(logits=logits)
+        action = policy_dist.sample()
+        # Action probabilities for calculating the adapted soft-Q loss
+        action_probs = policy_dist.probs
+        log_pi = F.log_softmax(action_probs, dim=1)
+        return action, log_pi, action_probs
+
+
+if __name__ == "__main__":
+    args = parse_args()
+    run_name = f"{args.env_id}__{args.exp_name}__{args.seed}__{int(time.time())}"
+    if args.track:
+        import wandb
+
+        wandb.init(
+            project=args.wandb_project_name,
+            entity=args.wandb_entity,
+            sync_tensorboard=True,
+            config=vars(args),
+            name=run_name,
+            monitor_gym=True,
+            save_code=True,
+        )
+    writer = SummaryWriter(f"runs/{run_name}")
+    writer.add_text(
+        "hyperparameters",
+        "|param|value|\n|-|-|\n%s" % ("\n".join([f"|{key}|{value}|" for key, value in vars(args).items()])),
+    )
+
+    # Don't utilize full CPU or else I might get warned
+    # TODO: Must change back before merge
+    torch.set_num_threads(1)
+
+    # TRY NOT TO MODIFY: seeding
+    random.seed(args.seed)
+    np.random.seed(args.seed)
+    torch.manual_seed(args.seed)
+    torch.backends.cudnn.deterministic = args.torch_deterministic
+
+    # TODO: Change back before merge
+    device = torch.device("cuda:0" if torch.cuda.is_available() and args.cuda else "cpu")
+
+    # env setup
+    envs = gym.vector.SyncVectorEnv([make_env(args.env_id, args.seed, 0, args.capture_video, run_name)])
+    assert isinstance(envs.single_action_space, gym.spaces.Discrete), "only discrete action space is supported"
+
+    actor = network_init(Actor(envs)).to(device)
+    qf1 = network_init(SoftQNetwork(envs)).to(device)
+    qf2 = network_init(SoftQNetwork(envs)).to(device)
+
+    qf1_target = SoftQNetwork(envs).to(device)
+    qf2_target = SoftQNetwork(envs).to(device)
+    qf1_target.load_state_dict(qf1.state_dict())
+    qf2_target.load_state_dict(qf2.state_dict())
+    q_optimizer = optim.Adam(list(qf1.parameters()) + list(qf2.parameters()), lr=args.q_lr, eps=1e-4)
+    actor_optimizer = optim.Adam(list(actor.parameters()), lr=args.policy_lr, eps=1e-4)
+
+    # Automatic entropy tuning
+    if args.autotune:
+        target_entropy = -0.98 * torch.log(torch.tensor(envs.single_action_space.n, dtype=float))
```

Comment:
What's the reference for this implementation ?

Reply:
Yeah, I think max entropy should be -log(1 / |A|) (from entropy definition for discrete random variables). For example, that's what I have in my [implementation](https://github.com/Howuhh/hse_reinforcement_learning/blob/main/extra_algo/sac_discrete.ipynb) and [others](https://github.com/ku2482/sac-discrete.pytorch/blob/master/sacd/agent/sacd.py#L47)

Reply:
You're both correct. I messed this up when going from numpy to torch